share/wake/lib/system/remote_cache_runner.wake

```diff
@@ -216,79 +213,72 @@ export def mkRemoteCacheRunner (rscApi: RemoteCacheApi) (hashFn: Result RunnerIn
             | findFail
             | addErrorContext "rsc: Failed to create a symlink"
 
-        def _ = virtual job stdout stderr status runtime cputime mem ibytes obytes
+        def _ = primJobVirtual job stdout stderr predict
 
         Pass (RunnerOutput inputs outputs predict)
 
-    def doit job runnerInput = match runnerInput
-        Fail e ->
-            def _ = badlaunch job e
+    def run (job: Job) (input: RunnerInput): Result RunnerOutput Error =
+        def label = input.getRunnerInputLabel
 
-            Fail e
-        Pass input ->
-            def label = input.getRunnerInputLabel
+        require Pass hashKey = hashFn input
 
-            require Pass hashKey = hashFn runnerInput
+        # If pulling from the cache is not enabled don't bother searching.
+        require True = rscApi.getRemoteCacheApiCanPull
+        else runJobAndUpload job input hashKey
 
-            # If pulling from the cache is not enabled don't bother searching.
-            require True = rscApi.getRemoteCacheApiCanPull
-            else runJobAndUpload job input hashKey
+        # -------------------------------------
+        # --- Search the cache for the job  ---
+        # -------------------------------------
 
-            # -------------------------------------
-            # --- Search the cache for the job  ---
-            # -------------------------------------
+        # Search the cache for a match
+        def requestTask =
+            rscApi
+            | rscApiFindMatchingJob (mkSearchRequest input hashKey)
 
-            # Search the cache for a match
-            def requestTask =
-                rscApi
-                | rscApiFindMatchingJob (mkSearchRequest input hashKey)
+        require Pass response = requestTask
+        else
+            require Fail err = requestTask
+            else unreachable "Result must be either Pass or Fail"
```

Comment:
I know this is just deindenting the existing code rather than being something new, but since you're touching this already, and especially given how complex both branches are, mind refactoring it into a match with helper functions?
```suggestion
        match requestTask
            Pass response -> runMatchedJob response
            Fail err -> runUncachedJob err
```

Reply:
The two arms here aren't Hit vs Miss, its "Got a valid server response" vs "Got an invalid server response" and the two extracted functions in the match would be kind of silly imo. TBH this is something I **_really_** want `when` for

```python
    # If a failure occurs making the request to the server log it and exit
    when Fail err = requestTask
        # Always leave a breadcrumb since this should be a rare error.
        def _ = breadcrumb "{label}: Failed to search for job in the cache"

        # Leave detailed info if debugging is enabled
        def _ =
            require True = shouldDebugRemoteCache Unit

            def _ =
                writeTempFile "remote.cache.lookup.fail" "label: {input.getRunnerInputLabel}\nerror: {err | format}"

            True

        # This job isn't getting cached. That's probably preferable since the server
        # request failed but good to keep in mind.
        baseDoIt job (Pass input)

    require Pass response = requestTask

    # If a match was found use it
    require NoMatch = response
    else
        ... rehydrate the job from the cache ...

    ... run and cache the job ...
```

Another also kind of weird way to do this would be a call to `rmapFail`. I don't really see this branch as "do one of two things" instead its "do one thing, or on failure recover and log"

Reply:
Yeah, `when` might be nice for this.  My view is just that (A) the `else` case here is large enough that it's beginning to get difficult to tell at a glance where it ends, and (B) that the series of two immediate `require`s and an `unreachable` looks a bit silly when the ultimate goal (connotations aside) is the same as an all-constructors `match`.  I think it probably comes down to how we individually think of `match`/`require` -- to my eye, `match` doesn't imply that its cases are all equally prominent, so having the Fail case just be recovery/logging is entirely natural; meanwhile I tend to look at `require` as something which should "return control to the enclosing function" (despite that being *entirely* the incorrect summary of what's behind the scenes) as quickly as possible, with neither the RHS of the `=` nor the `else` clause ideally containing any significant code.

Not a critical issue either way, though.